src/state_machine.zig

```diff
@@ -915,34 +917,64 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
             assert(dr_mut.timestamp == dr_immut.timestamp);
             assert(cr_mut.timestamp == cr_immut.timestamp);
 
+            const amount = amount: {
+                var amount = t.amount;
+                if (t.flags.balancing_debit or t.flags.balancing_credit) {
+                    if (amount == 0) amount = std.math.maxInt(u64);
+                }
+
+                if (t.flags.balancing_debit) {
+                    const dr_balance = dr_mut.debits_posted + dr_mut.debits_pending;
+                    const dr_limit = if (dr_immut.flags.debits_must_not_exceed_credits)
+                        dr_mut.credits_posted
+                    else
+                        std.math.maxInt(u64);
+                    amount = std.math.min(amount, dr_limit - dr_balance);
+                }
```

Comment:
If I'm reading this right, then I think we may have a bug here, when:

1. `dedits_must_not_exceed_crebits` is not set, and
2. `amount == 0` i.e. maxInt(u64), then
3. the credit balance (`credits_posted`) will be ignored completely, so that
4. a higher amount than expected will be debited.

Instead, the basic principle with these balancing transfers (setting the further "at most" feature aside), is that the calculated amount should never exceed the net credit balance, of credits less debits (taking a conservative view of credits so as to exclude pending credits, and taking a conservative view of the balancing side so as to include pending debits).

With the "at most" cap in place, it should never exceed the max cap, but the net balance should nevertheless also be enforced. Does that make sense?

Otherwise, it would be great to explore if we can DRY up both these `balancing_debit/credit` branches, by using a single inlined function into which we pass the respective amounts and flag, twice over but with the opposite values each time, since the code is perfectly symmetrical. This could also help to reduce the bulk of this block scope.

I'm also still nervous that we rely on other code further below to check that the final balancing amount is not 0. I think we should have explicit checks closer in "space" to where we do these balancing calculations, to rule out invalid values earlier. When related checks are spread out in the code base, it's almost like another form of TOCTOU, but instead of "time", the human-confusing part is "space".

For example, what happens if the account is already balanced? It seems like we should have an error code specifically in this context of balancing transfers?

Reply:
I am reluctant to DRY this up with a helper function — I think "hiding" this core logic with an indirection would negate the benefit of cleanup up the block.

Is this better?

```
            const amount = amount: {
                var amount = t.amount;
                if (t.flags.balancing_debit or t.flags.balancing_credit) {
                    if (amount == 0) amount = std.math.maxInt(u64);
                } else {
                    assert(amount != 0);
                }

                if (t.flags.balancing_debit) {
                    const dr_balance = dr_mut.debits_posted + dr_mut.debits_pending;
                    amount = std.math.min(amount, dr_mut.credits_posted -| dr_balance);
                    if (amount == 0) return .already_balanced_debit_account;
                }

                if (t.flags.balancing_credit) {
                    const cr_balance = cr_mut.credits_posted + cr_mut.credits_pending;
                    amount = std.math.min(amount, cr_mut.debits_posted -| cr_balance);
                    if (amount == 0) return .already_balanced_credit_account;
                }
                break :amount amount;
            };
```